### PR TITLE
Remove old remote lxd client

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

Remove the remote from the lxd tool client. This should mean that
all connections to the remote server should be through the new lxd
container and not through the old tool/lxdclient.

## QA steps

- Test that you can connect via the normal localhost setups.
- Testing a connection to a remote setup is possible (with appropriate 
  certs)

## Documentation changes

- No changes externally.

## Bug reference

https://trello.com/c/uKWwlPMh/314-remove-old-lxd-client-and-use-new-lxd-server-directly-in-lxd-provider

### CODE REVIEW

File: container/lxd/server.go [hunk 1]
```diff
@@ -63,6 +63,16 @@ func NewLocalServer() (*Server, error) {
 	return svr, errors.Trace(err)
 }
 
+// NewRemoteServer returns a Server based on a remote connection.
+func NewRemoteServer(remote RemoteServer) (*Server, error) {
+	cSvr, err := ConnectRemote(remote)
```
Comment (manadart) [inline container/lxd/server.go]: Here, where we know for sure that we are connecting to the remove and then using the connection to create a server, we can prevent a double request by doing:
```
remote.SkipGetServer = true
```
As this is executing first thing in NewServer.

File: container/lxd/server.go [hunk 2]
```diff
@@ -63,6 +63,16 @@ func NewLocalServer() (*Server, error) {
 	return svr, errors.Trace(err)
 }
 
+// NewRemoteServer returns a Server based on a remote connection.
+func NewRemoteServer(remote RemoteServer) (*Server, error) {
```
Comment (manadart) [inline container/lxd/server.go]: The nomenclature might get a little confusing here. We have ```NewRemoteServer``` that is not returning a ```RemoteServer``` instance as might be expected.

Maybe a better name for ```RemoteServer``` is something like ```ServerSpec```. And we could abstract away dealing with the upstream ```ConnectionArgs``` type by have a ```NewServerSpec``` method that takes name, host, protocol, client cert, server cert.

File: provider/lxd/environ_raw.go [hunk 3]
```diff
@@ -67,8 +65,19 @@ func newRawProvider(spec environs.CloudSpec, local bool) (*rawProvider, error) {
 		prov, err := newLocalRawProvider()
```
Comment (manadart) [inline provider/lxd/environ_raw.go]: As we strangle this out, this is where we would move the default network and storage checks from the tools/lxdclient client.

File: provider/lxd/environ_raw.go [hunk 4]
```diff
@@ -77,40 +86,13 @@ func newLocalRawProvider() (*rawProvider, error) {
...
-	config, err := getRemoteConfig(spec)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	raw, err := newRawProviderFromConfig(*config)
-	return raw, errors.Trace(err)
-}
-
 func newRawProviderFromConfig(config jujulxdclient.Config) (*rawProvider, error) {
 	client, err := jujulxdclient.Connect(config, true)
```
Comment (manadart) [inline provider/lxd/environ_raw.go]: If we remove config from tools/lxdclient as well, we can replace the config argument with RemoteServer (ServerSpec?).

Then there is only a little to do before Connect itself is no longer needed and we use Server directly.

File: container/lxd/server.go (hunk 2 above)
Comment (SimonRichardson) [reply container/lxd/server.go]: Agreed, in fact this is what was confusing with the old `tools/lxdclient` and hard to prise apart.

File: container/lxd/connection.go [hunk 5]
```diff
@@ -21,17 +21,57 @@ const (
...
+type ServerSpec struct {
+	Name           string
+	Host           string
+	Protocol       Protocol
+	connectionArgs *lxd.ConnectionArgs
+}
+
+// MakeServerSpec creates a ServerSpec with default values where needed.
+// It also ensures the HTTPS for the host implicitly
+func MakeServerSpec(host, serverCert string, clientCert *Certificate) ServerSpec {
```
Comment (manadart) [inline container/lxd/connection.go]: I'd call it ```NewServerSpec[option]``` in keeping with convention.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**